drivers/flash/flash_stm32_ospi.c

```diff
@@ -0,0 +1,1422 @@
+
+	cmd->NbData = size;
+
+	dev_data->cmd_status = 0;
+
+	/* in OPI/STR the 3-byte AddressSize is not supported by the NOR flash */
+	if ((dev_cfg->data_mode == OSPI_OPI_MODE) && (cmd->AddressSize != HAL_OSPI_ADDRESS_32_BITS)) {
+		LOG_ERR("OSPI wr in OPI/STR mode is for 32bit address only");
+		return -EIO;
+	}
+
+	hal_ret = HAL_OSPI_Command(&dev_data->hospi, cmd, HAL_OSPI_TIMEOUT_DEFAULT_VALUE);
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to send OSPI instruction", hal_ret);
+		return -EIO;
+	}
+
+	hal_ret = HAL_OSPI_Transmit_IT(&dev_data->hospi, (uint8_t *)data);
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to read data", hal_ret);
+		return -EIO;
+	}
+
+	k_sem_take(&dev_data->sync, K_FOREVER);
+
+	return dev_data->cmd_status;
+}
+
+/*
+ * Gives a OSPI_RegularCmdTypeDef with all parameters set
+ * except Intruction, Address, DummyCycles, NbData
+ */
+static OSPI_RegularCmdTypeDef ospi_prepare_cmd(uint8_t transfer_mode, uint8_t transfer_rate)
+{
+
+	OSPI_RegularCmdTypeDef cmd_tmp = {
+		.OperationType = HAL_OSPI_OPTYPE_COMMON_CFG,
+		.FlashId = HAL_OSPI_FLASH_ID_1,
+		.InstructionMode = ((transfer_mode == OSPI_SPI_MODE)
+				? HAL_OSPI_INSTRUCTION_1_LINE
+				: HAL_OSPI_INSTRUCTION_8_LINES),
+		.InstructionSize = ((transfer_mode == OSPI_SPI_MODE)
+				? HAL_OSPI_INSTRUCTION_8_BITS
+				: HAL_OSPI_INSTRUCTION_16_BITS),
+		.InstructionDtrMode = ((transfer_rate == OSPI_DTR_TRANSFER)
+				? HAL_OSPI_INSTRUCTION_DTR_ENABLE
+				: HAL_OSPI_INSTRUCTION_DTR_DISABLE),
+		.AddressMode = ((transfer_mode == OSPI_SPI_MODE)
+				? HAL_OSPI_ADDRESS_1_LINE
+				: HAL_OSPI_ADDRESS_8_LINES),
+		.AddressDtrMode = ((transfer_rate == OSPI_DTR_TRANSFER)
+				? HAL_OSPI_ADDRESS_DTR_ENABLE
+				: HAL_OSPI_ADDRESS_DTR_DISABLE),
+		.AddressSize = HAL_OSPI_ADDRESS_32_BITS,
+		.AlternateBytesMode = HAL_OSPI_ALTERNATE_BYTES_NONE,
+		.DataMode = ((transfer_mode == OSPI_SPI_MODE)
+				? HAL_OSPI_DATA_1_LINE
+				: HAL_OSPI_DATA_8_LINES),
+		.DataDtrMode = ((transfer_rate == OSPI_DTR_TRANSFER)
+				? HAL_OSPI_DATA_DTR_ENABLE
+				: HAL_OSPI_DATA_DTR_DISABLE),
+		.DQSMode = (transfer_rate == OSPI_DTR_TRANSFER)
+				? HAL_OSPI_DQS_ENABLE
+				: HAL_OSPI_DQS_DISABLE,
+		.SIOOMode = HAL_OSPI_SIOO_INST_EVERY_CMD,
+	};
+
+	return cmd_tmp;
+}
+
+/*
+ * Read Serial Flash Discovery Parameter :
+ * perform a read access over SPI bus for SDFP (DataMode is already set)
+ * or get it from the sdfp table (in the DTS)
+ */
+static int ospi_read_sfdp(const struct device *dev, off_t addr, uint8_t *data,
+			  size_t size)
+{
+	const struct flash_stm32_ospi_config *dev_cfg = dev->config;
+	struct flash_stm32_ospi_data *dev_data = dev->data;
+
+#if DT_NODE_HAS_PROP(DT_INST(0, st_stm32_ospi_nor), sfdp_bfp)
+	/* simulate the SDFP */
+	ARG_UNUSED(addr); /* addr is 0 */
+
+	for (uint8_t i_ind = 0; i_ind < ARRAY_SIZE(dev_cfg->sfdp_bfp); i_ind++) {
+		*(data + i_ind) = dev_cfg->sfdp_bfp[i_ind];
+	}
+#else /* sfdp_bfp */
+
+	OSPI_RegularCmdTypeDef cmd = {
```

Comment:
Why not calling `ospi_prepare_cmd()` here ?

Reply:
yes, including the ospi_prepare_cmd here